spec.html

```diff
@@ -24176,16 +24390,392 @@ <h1>Runtime Semantics: Evaluation</h1>
+
+      DecoratorMemberExpression[Yield, Await] :
+        IdentifierReference[?Yield, ?Await]
+        DecoratorMemberExpression[?Yield, ?Await] `.` IdentifierName
+
+      DecoratorParenthesizedExpression[Yield, Await] :
+        `(` Expression[+In, ?Yield, ?Await] `)`
+
+      DecoratorCallExpression[Yield, Await] :
+        DecoratorMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
+        DecoratorParenthesizedExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
+    </emu-grammar>
+
+    <emu-clause id="sec-decoratorevaluation" type="sdo">
+      <h1>Runtime Semantics: DecoratorEvaluation</h1>
+      <emu-grammar>Decorator : `@` DecoratorMemberExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |MemberExpression| that is covered by |DecoratorMemberExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
+        1. Return ? GetValue(_ref_).
+      </emu-alg>
+      <emu-grammar>Decorator : `@` DecoratorCallExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |CallMemberExpression| that is covered by |DecoratorCallExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
+        1. Return ? GetValue(_ref_).
+      </emu-alg>
+      <emu-grammar>Decorator : `@` DecoratorParenthesizedExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |MemberExpression| that is covered by |DecoratorParenthesizedExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
+        1. Return ? GetValue(_ref_).
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-decoratorlistevaluation" type="sdo">
+      <h1>Runtime Semantics: DecoratorListEvaluation</h1>
+      <emu-grammar>DecoratorList : DecoratorList? Decorator</emu-grammar>
+      <emu-alg>
+        1. If |DecoratorList| is present, then
+          1. Let _leftValue_ be ? DecoratorListEvaluation of |DecoratorList|.
+        1. Else,
+          1. Let _leftValue_ be a new empty List.
+        1. Let _rightValue_ be ? DecoratorEvaluation of |Decorator|.
+        1. Prepend _rightValue_ to _leftValue_.
+        1. Return _leftValue_.
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-createdecoratoraccessobject" type="abstract operation">
+      <h1>
+        CreateDecoratorAccessObject (
+          _kind_: ~field~, ~accessor~, ~method~, ~getter~, or ~setter~,
+          _name_: a Private Name,
+        ): an Object
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>Returns an object with appropriate accessor functions for the private value, for use in a decorator context object.</dd>
+      </dl>
+      <emu-alg>
+        1. Let _accessObj_ be OrdinaryObjectCreate(%Object.prototype%).
+        1. If _kind_ is ~field~, ~accessor~, ~method~, or ~getter~, then
+          1. Let _getterClosure_ be a new Abstract Closure with no parameters that captures _name_ and performs the following steps when called:
+            1. Let _o_ be the *this* value.
```

Comment:
**this** will be the `accessObj`. Shouldn't it always bond to the class instance?

Reply:
when it called as `accessObj.get()`

Reply:
It should not, the `accessObj` is created once per class, since decoration only occurs once per class. The user must call the function with `this` being the instance when actually using the function. This is an unintuitive but advanced API, and decorator authors are used to doing this already since decoration only occurs once on the prototype with all iterations of the proposal.

Reply:
Getters, like all prototype methods, should be borrowable - both .call-able on other objects, and also should work with inheritance, all of which requires dynamically looking up the receiver.

Reply:
so the correct usage of this API should be `accessObj.get.call(this)`? This is not intuitive...

Reply:
Correct, this is a technique which decorator authors use often. For instance, when decorating a method, you must call the original method this way:

```js
addOne(original) {
  return function() {
    return original.call(this) + 1;
  }
}
```

I agree that it is not intuitive, but metaprogramming rarely is in my experience. Importantly, this is only complicated for decorator _authors_. Decorator users, who are far more common in general, never have to work with these unintuitive APIs.

Reply:
it can be more intuitive, like

```js
accessObj.get(classInstance)
```

and `accessObj.get()` default binds to the current instance

Reply:
that would not be intuitive, because it's not how anything in the language behaves.

Reply:
hmm, I don't expect writing a decorator needs to know `receivers`/`this` so clearly to know they should write `access.get.call(instance)`.

I think getting `this` from the first parameter, in this case, isn't too bad.

Reply:
There is also a purely syntactic way to perform this delegation when needed:

```js
function addOne(method) {
  return {
    __proto__: { method },
    method() {
      return super.method() + 1;
    }
  }.method;
}
```

(If `::` or similar proposals succeed, they will cover this scenario without the awkward [[HomeObject]] boilerplate.)